src/content/post/20161122_drupal_application_bottlenecks_avoid_with_profiling.md

```diff
@@ -0,0 +1,160 @@
+
+1. Check caches
+    1. Check Page cache, block cache and view cache
+2. Aggregate and minify CSS and js
+3. Deactivate development component
+    1. Modules like _devel_, _update_status_, _performance_ are really performances hogs.
+4. Deactivate _cron_ automation
+    1. Be sure you are running batches and asynchronous tasks during low-traffic hours, wherever possible. Remember to demand critical tasks or heavy tasks to external workers.
+5. Deactivate database logger, switching to _syslog_
+6. Check images size
+    1. Use opportune _image_styles_
+7. Check retrieval process of external resources (feed, streams, etc.). I mean that we need to reply to questions like `How many time the application need to retrieve last Company's tweets?`, `How many time was spent to collect weather info to populate this badge?`, `It's normal that before to load the page we collect info related to the last match results?`
+
+### Detection of application's bottleneck
+
+Once we went through all the list above, your server will probably sigh in relief. This doesn't mean our work is done: we need to use right tools (especially if we lack specific references to slow page) to find the cause of th high server load. Before doing anything else we need to collect data about what's going under the hood, so we can improve in the right direction. Yes, I'm talking about [profiling](https://en.wikipedia.org/wiki/Profiling_(computer_programming)).
+
+#### Application side
+
+On Drupal 7, we can use the [Performance Logging](https://www.drupal.org/project/performance) module to profile time spent to generate pages, memory consumption, and how many queries fire during page load. We only need to navigate our site to populate the data table. On Drupal 8 the great _Web Profiler_ module, recently merged into [Devel](https://www.drupal.org/project/devel) project is the best option.
+
+#### Environment side
+
+The environment (local but also production), with the right tools can become our better friend to help us to find reasons of application's bottlenecks. There is a lot of profiler, but an excellent choice is [Blackfire](http://blackfire.io/): free, flexible and very light (his agent can be installed on the production servers with a really really thin overhead). The feature which makes me love Blackfire are his profiler result: a simple and clear diagram of entire code execution flow from the request response to the final output: all function calls will be analyzed and counted to produce a detailed report that contain a lot of useful data like function's time execution, memory consumption, etc. Finally all this data will be collected and the results will be showed to us related to the global status: a function that takes three seconds to produce an output, can be a lower problem related to another function that takes only 0.5 second but it's called ninety time.
+Obviously, remember to puts on the mysql slow query log.
+
+#### An example:
```

Comment:
I'd keep this up one level and retitle like this:

`### A practical profiling example